File: algorithms/2_medium/W/00874_walking_robot_simulation/hashset.cpp
constexpr size_t dir_len = 4;
constexpr int directions [dir_len][2] = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};

class Solution {
    struct PairHash {
        std::size_t operator()(const std::pair<int, int>& p) const {
            static const auto hash_func = std::hash<int>();
            const auto h1 = hash_func(p.first);
            const auto h2 = hash_func(p.second);
            return h1 ^ (h2 << 1);
        }
    };
public:
    int robotSim(const vector<int>& commands, const vector<vector<int>>& obstacles) {
        assert(all_of(obstacles.cbegin(), obstacles.cend(), [](const auto& a){
            return std::ssize(a) == 2;
        }));
        auto hashset = unordered_set<pair<int, int>, PairHash>();
        for (const auto& o : obstacles) {
            hashset.insert(make_pair(o[0], o[1]));
        }
        int x = 0;
        int y = 0;
        size_t dir_idx = 0;
        int max = 0;
        for (const int c : commands) {
            switch (c) {
                case -2:
                    dir_idx = (dir_idx + dir_len - 1) % dir_len;
                    break;
                case -1:
                    dir_idx = (dir_idx + 1) % dir_len;
                    break;
                default: {
                    if (c < 1 || c > 9) {
                        cerr << "invalid input";
                        exit(1);
                    }
                    const auto [dx, dy] = directions[dir_idx];
                    for (int steps = 1; steps <= c; steps += 1) {
                        if (hashset.contains(make_pair(x + dx, y + dy))) {
                            break;
                        }
                        x += dx;
                        y += dy;
                        max = std::max(max, x * x + y * y);
                    }
                    break;
                }
            }
        }

        return max;
    }
};

File: algorithms/2_medium/W/00874_walking_robot_simulation/hashset.rs
use std::collections::HashSet;

impl Solution {
    const DIR_LEN: usize = 4;
    const DIRECTIONS: [(i32, i32); Self::DIR_LEN] = [(0, 1), (1, 0), (0, -1), (-1, 0)];

    pub fn robot_sim(commands: Vec<i32>, obstacles: Vec<Vec<i32>>) -> i32 {
        assert!(obstacles.iter().all(|o| o.len() == 2));
        let mut set: HashSet<(i32, i32)> = obstacles.into_iter().map(|o|(o[0], o[1])).collect();
        let (mut x, mut y) = (0_i32, 0_i32);
        let mut dir_idx = 0_usize;
        let mut max = 0;
        for c in commands {
            match c {
                -2 => dir_idx = (dir_idx + Self::DIR_LEN - 1) % Self::DIR_LEN,
                -1 => dir_idx = (dir_idx + 1) % Self::DIR_LEN,
                1..=9 => {
                    let (dx, dy) = Self::DIRECTIONS[dir_idx];
                    for _ in 1..=c {
                        if set.contains(&(x + dx, y + dy)) {
                            break;
                        }
                        x += dx;
                        y += dy;
                        max = max.max(x * x + y * y);
                    }
                },
                _ => unreachable!("invalid input"),
            }
        }

        max
    }
}
